docs/cn/2019-09-09-BR-key-rewrite-disscussion.md

```diff
@@ -0,0 +1,89 @@
+Last updated: 2019-09-12
+
+## BR 关于 Key rewrite 的讨论
+
+### Key rewirte 结构
+
+Key Rewrite 的目的有二：
+
+1. 为 BR 提供修改 Table ID 的功能，以支持恢复到 Schema Version 不同的集群
+2. 为 Lightning 提供添加前缀的功能，省略 Lightning <-> Importer 之间重复的数据传输
+
+一个 BR 的 SST 可能包含多个 Tables，所以要支持多条 Rewrite Rules 同时生效。SST 可能来自非 TiDB 系统，所以 Importer 不应该有 Key 编码格式的假设（不一定是 t«tid»_ 开头）。
+
+给 Importer / TiKV 参考的 Key Rewrite 数据结构建议如下：
+
+```protobuf
+message RewriteRule {
+	bytes old_prefix = 1;  // this can be empty for universal prefix insertion!
+	bytes new_prefix = 2;  // these are _not_ just an integer!
+}
+
+message RestoreRequest {
+	...
+	repeated RewriteRule rewrite_rules = N;
+	...
+}
+```
+
+正向替代一个 Key：
+
+```rust
+fn rewrite_key(rules: &[RewriteRule], key: &[u8]) -> Cow<[u8]> {
+    for rule in rules {
+        if key.starts_with(rule.old_prefix) {
+            return Cow::Owned(rule.new_prefix + key[rule.old_prefix.len()..])
+        }
+    }
+    Cow::Borrowed(key)
+}
+```
+
+反向还原一个 Key：
+
+```rust
+fn undo_rewrite_key(rules: &[RewriteRule], key: &[u8]) -> Cow<[u8]> {
+    for rule in rules {
+        if key.starts_with(rule.new_prefix) {
+            return Cow::Owned(rule.old_prefix + key[rule.new_prefix.len()..])
+        }
+    }
+    Cow::Borrowed(key)
+}
+```
+
+### Key Rewrite 对现在导入流程的影响
+
+现在无论是 BR 还是 Lightning，使用的都是同一套导入流程，如下：
+1. 把 **KV 对**写入到 RocksDB 实例（“**Engine File**”）来排序
+2. (Prepare) 遍历此 Engine file，
+   1. (Pre-split) 每当 K+V 的累积长度达 512 MB 时（可设置），对这个 **Range** 执行 PrepareRangeJob，如下。
```

Comment:
There are Ligthning only. Cc @3pointer

Reply:
This discussion is about tikv-importer, which would be replaced by the download & ingest API. That seems like, split region will be done at "client" side, both BR and lightning can have their own behavior.

As a result, should we replace description with current BR behavior?